org.adempiere.base.process/src/org/compiere/process/InOutGenerate.java

```diff
@@ -290,7 +291,17 @@ private String generate (PreparedStatement pstmt)
 							logInfo.append(" (set to 0)");
 						}
 						//	Adjust On Hand
-						onHand = onHand.subtract(unconfirmedShippedQty);
+						StringBuilder where3 = new StringBuilder ("EXISTS (SELECT * FROM M_InOut io WHERE io.M_InOut_ID=M_InOutLine.M_InOut_ID AND io.IsSOTrx = 'Y' AND io.DocStatus IN ('DR','IN','IP','WC')")
+							    .append	(" AND io.M_Warehouse_ID=").append(p_M_Warehouse_ID).append (")");;
+						iols = MInOutLine.getOfProduct(getCtx(), 
+							line.getM_Product_ID(), where3.toString(), get_TrxName());
+						for (int j = 0; j < iols.length; j++) 
+							// don't count lines of our m_shipment, they are already decreased in storage
+							if (m_shipment == null || iols[j].get_ValueAsInt("M_InOut_ID") != m_shipment.get_ID())
+								totalunconfirmedShippedQty = totalunconfirmedShippedQty.add(iols[j].getMovementQty());
+						logInfo = new StringBuilder("TotalUnconfirmed Qty=").append(totalunconfirmedShippedQty) 
```

Comment:
this is inefficient as you can simply do a sum of movementqty for all m_inout of same product, warehouse and id != current shipment id.

Reply:
You are absolutely correct. If I understood you corretly, this is closer to what you'd have in mind?

Reply:
yes, this is better but I think we need to verify that the generate sql is indeed efficient (i.e cost analysis through pgadmin).

Reply:
2 seq scan and that's going to be slow for big db, need new index.

Reply:
2 seq scan and that's going to be slow for big db, need new index.

Reply:
Hi @hengsin, I have to apologize, I made a mistake. For my test I used a development database containing next to no entries. After retrying the test on a larger database, it looked like this:
![grafik](https://user-images.githubusercontent.com/79985645/168984831-13f45514-719c-4487-9a79-4beec2aecf9c.png)

If you still see the need for an additional index, I'd like some hints on what these indexes should encompass. Thank you for your patience